File: 210_Course_Schedule_II/210_2.py
from typing import List, Dict

class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        if numCourses == 0:
            return []
        if not prerequisites:
            return list(reversed(range(numCourses)))

        valid = True
        stack = []
        graph = [[] for _ in range(numCourses)]
        visited = [0 for _ in range(numCourses)]
        for p in prerequisites:
            graph[p[1]].append(p[0])

        def dfs(index) -> None:
            nonlocal valid
            visited[index] = 1
            for v in graph[index]:
                if visited[v] == 0:
                    dfs(v)
                    if not valid:
                        return
                elif visited[v] == 1:
                    valid = False
                    return
            
            visited[index] = 2
            stack.append(index)

        for i in range(numCourses):
            if valid and visited[i] == 0:
                dfs(i)
        if not valid:
            return []
        return stack[::-1]

if __name__ == "__main__":
    print(Solution().findOrder(4, [[1,0],[2,0],[3,1],[3,2]]))

File: 210_Course_Schedule_II/210_2.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> res;
        if (numCourses == 0) return vector<int>(0);
        if (prerequisites.empty()) {
            for (int i = 0; i < numCourses; i++) {
                res.push_back(i);
            }
            return res;
        }

        vector<vector<int>> graph(numCourses, vector<int>());
        for (vector<int> &p : prerequisites) {
            graph[p[1]].push_back(p[0]);
        }

        bool valid = true;
        vector<int> visited(numCourses, 0);
        vector<int> stack;
        for (int i = 0; i < numCourses && valid; i++) {
            if (visited[i] == 0) {
                dfs(i, graph, visited, stack, valid);
            }
        }

        if (!valid) {
            return vector<int>();
        }
        reverse(stack.begin(), stack.end());
        return stack;
    }

    void dfs(int index, vector<vector<int>> &graph,
             vector<int> &visited, vector<int> &stack, bool &valid) {
        visited[index] = 1;
        for (const int &v : graph[index]) {
            if (visited[v] == 0) {
                dfs(v, graph, visited, stack, valid);
                if (!valid) return;
            }
            else if (visited[v] == 1) {
                valid = false;
                return;
            }
        }
        visited[index] = 2;
        stack.push_back(index);
    }
};
